jax/_src/lax/lax.py

```diff
@@ -2810,11 +2830,25 @@ def _conv_general_dilated_shape_rule(
                                batch_group_count)
   return tuple(np.take(out_trans, np.argsort(out_perm)))  # type: ignore[arg-type]
 
+def _validate_preferred_elemet_type(input_dtype, preferred_element_type):
+  if dtypes.issubdtype(input_dtype, np.integer) and not dtypes.issubdtype(preferred_element_type, np.integer):
+    raise TypeError("`preferred_element_type` and the original type must both be integral or both be floating point.")
```

Comment:
It looks like this doesn't catch the case when `input_dtype` is a float and `preferred_element_type` is an integer.

Do we need a similar check for complex?

Reply:
> It looks like this doesn't catch the case when input_dtype is a float and preferred_element_type is an integer.

Good catch! This checking was just copied from `lax.dot` The interpreter throws a sensible error message, but I agree it's better to catch it already here.